### PR TITLE
Added Human-Readable ABI format

### PR DESCRIPTION
**Pull request by andresceballosm**
**Type of PR:**

- [ ] Bugfix
- [X] Feature
- [ ] Documentation
- [ ] Other

**Required reviews:**

- [ ] 1
- [ ] 2
- [X] 3

**What this does:**
> Added Human-Readable ABI format

**Issues fixed/closed:**
> - Fixes https://github.com/nucypher/taco-web/issues/370

### CODE REVIEW

File: packages/taco/src/conditions/base/contract.ts [hunk 1]
```diff
@@ -70,6 +70,31 @@ const functionAbiSchema = z
...
+    /^function\s+/,
+    '',
+  );
+  const fragment = ethers.utils.FunctionFragment.from(
+    abiWithoutFunctionKeyword,
+  );
+  const jsonAbi = JSON.parse(fragment.format(ethers.utils.FormatTypes.json));
+
+  delete jsonAbi.constant;
+  delete jsonAbi.payable;
```
Comment (piotr-roslaniec) [inline packages/taco/src/conditions/base/contract.ts:84]: Why do we delete these fields? Can we use destructuring and only pick relevant fields instead?

File: packages/taco/src/conditions/base/contract.ts [hunk 2]
```diff
@@ -70,6 +70,31 @@ const functionAbiSchema = z
...
+
+  jsonAbi.inputs = jsonAbi.inputs.map((input: any) => ({
+    ...input,
+    internalType: input.type,
+  }));
+  jsonAbi.outputs = jsonAbi.outputs.map((output: any) => ({
+    ...output,
+    internalType: output.type,
+  }));
```
Comment (piotr-roslaniec) [inline packages/taco/src/conditions/base/contract.ts:93]: Why do we need to map `type` to `internalType`? To be consistent with JSON ABI schema?

File: packages/taco/src/conditions/base/contract.ts [hunk 3]
```diff
@@ -82,7 +107,27 @@ export const contractConditionSchema = rpcConditionSchema
...
+                return true;
+              } catch (e) {
+                return false;
+              }
+            },
+            {
+              message: 'Invalid Human-Readable ABI format',
+            },
+          )
+          .transform(toJsonAbiFormat),
```
Comment (piotr-roslaniec) [inline packages/taco/src/conditions/base/contract.ts:128]: Let's move this variable into a separate declaration, just like `functionAbiSchema`

File: packages/taco/src/conditions/base/contract.ts [hunk 4]
```diff
@@ -98,9 +143,15 @@ export const contractConditionSchema = rpcConditionSchema
   );
 
 export type ContractConditionProps = z.infer<typeof contractConditionSchema>;
+interface ContractConditionHumanReadableAbi extends ContractConditionProps {
+  functionAbi: string;
+}
```
Comment (piotr-roslaniec) [inline packages/taco/src/conditions/base/contract.ts:148]: Why not infer it using `z.infer`?

File: packages/taco/src/conditions/base/contract.ts [hunk 5]
```diff
@@ -98,9 +143,15 @@ export const contractConditionSchema = rpcConditionSchema
...
 
 export class ContractCondition extends Condition {
-  constructor(value: OmitConditionType<ContractConditionProps>) {
+  constructor(value: OmitConditionType<ContractConditionHumanReadableAbi | ContractConditionProps>) {
+    if (typeof value.functionAbi === 'string') {
+      value.functionAbi = toJsonAbiFormat(value.functionAbi);
+    }
```
Comment (piotr-roslaniec) [inline packages/taco/src/conditions/base/contract.ts:154]: Why do we need to update this constructor and add a type invariant? Can we update `ContractConditionProps` (`z.infer`) instead?

File: packages/taco/test/conditions/base/contract.test.ts [hunk 6]
```diff
@@ -309,6 +320,40 @@ describe('supports custom function abi', () => {
...
+    expect(result.error).toBeDefined();
+    expect(result.data).toBeUndefined();
+    expect(result.error?.format()).toMatchObject({
+      functionAbi: {
+        _errors: ['Invalid Human-Readable ABI format'],
+      },
+    });
+  });
+
+  it('converts human-readable ABI to JSON ABI format', () => {
```
Comment (piotr-roslaniec) [inline packages/taco/test/conditions/base/contract.test.ts:339]: You can test the low-level behavior of `humanReadableAbiSchema` here

File: packages/taco/src/conditions/base/contract.ts (hunk 2 above)
Comment (andresceballosm) [reply packages/taco/src/conditions/base/contract.ts:109]: Yes!

**Review by theref — APPROVED** [review]

**Review by piotr-roslaniec — APPROVED** [review]
LGTM, thanks for contributing

**Review by vzotova — APPROVED** [review]

**Pull request merged**